---
title: "developing in epidemiology"
author: "Thijmen"
output: bookdown::html_document2
---

## Creating an epidemiology map

With the infrastructure we have created to make it easier for people to connect, we have also made the infrastructure for diseases to spread. Many of these diseases cause allot of trouble and can weaken whole societies, just take a look at SARS-COV-19. To figure out how these diseases spread is essential to figure out how a pathogen behaves. Figuring this out gives us the possibility to take precautions so it does not happen again, as can be seen by this report from the [CDC.](https://www.cdc.gov/eis/downloads/epidemiology-factsheet.pdf)        
In my future i want to be able to process epidemiological data to figure out where a disease originated, how it spread to different people and make this data easily readable for non data scientists with the help of shiny.       
To learn these skills i have made a small plan for a few steps i want to go through:        
1.    Find a suitable outbreak with available data.       
2.    Process the data to get simple phylogenetic trees.        
3.    Add regions to these trees.       
4.    Put these trees over a geographical map to see location.        
5.    Make these geopgraphicals maps interactive for easy visibility.       
<br>        

### Finding an outbreak

To be able and actually to create phylogenetic trees i need to have some data available to me.        
I dont want a massive amount of data like from the covid pandemic because this would be too much data to process while this is only for my own learning purpose.        
the NCBI has a very detailed database of different virus variation, for this exercise i will be using the data from the MERS-Cov outbreak (2013-2019).       
If you want to download this data for yourself go to [THIS SITE](https://www.ncbi.nlm.nih.gov/genome/viruses/variation/) and follow the following steps:        
1. Go to "MERS coronavirus"
2. Select nucleotide sequence type.       
3. Select Human host.        
4. Select S genome region.       
5. Click "additional filters", type "complete cds" and switch to definition lines.        
6. Click "Add query".        
7. Click "Show results". (this will redirect you to a new screen)       
8. Make sure only the S genome regions are selected!        
9. Click "Customize label", make sure the label only contains "\{accession\}". (this makes it easier to see later on)        
10. Click "Download" with as download option "Nucleotide (FASTA)".        
11. Click "Download" but now with as download option "Result set (CSV)". (This is the metadata)       
<br>        
You now have 2 files:       
1. The nucleotide sequence of all samples taken from human hosts with the accession number.       
2. The metadata that shows all the information for each accession number.       
<br>        
I will only be working with these 2 datafiles.        

### Creating Phylogenetic trees

For learning how to create these phylogenetic trees I have taken inspiration from a small tutorial on how to create Phylogenetic trees. You can find this tutorial [Here](https://fuzzyatelin.github.io/bioanth-stats/module-24/module-24.html).       


```{r}
# we need to load a few packages first
#BiocManager::install("msa")
library(msa)
library(tidyverse)
library(tidyr)
library(here)
library(adegenet)
library(ape)
#BiocManager::install("ggtree")
library(ggtree)
#BiocManager::install("treeio")
library(treeio)

# now we need to load in the nucleotides
dna <- readDNAStringSet(here("data/MERS_nucleotides.fa"))
dna

# and the metadata
metadata <- read.csv(here("data/MERS_annotation.csv"))
metadata %>% head(10) %>% knitr::kable()

# we only want the years from the metadata so lets extract those
metadata <- metadata %>% separate(collection_date, into = "collection_date", sep = 4)
metadata$collection_date <- as.numeric(metadata$collection_date)


# the DNA is loaded in nicely with the exact amount of downloaded sequences.
# there is no multiple sequence alignment necessary for these sequences, this is because they all come from the same part of the DNA.

dna <- as.DNAbin(dna) # sequence has to be the correct class

# to create a phylogenetic tree we need to calculate the distance between all the sequences
# there are many different models to choose from, and it was not easy to choose out of all the options because there are allot of upsides and allot of downsides to all the possible methods. for this time i have chosen to use the Tamaru and Nei methode (TN93). Because they take into account that its different to swap from A-G then C-T and vice versa.

dna_distance <- dist.dna(dna, model = "TN93")

# lets start by making a very simple neighbour joining phylogenetic tree.

nj_tree <- bionj(dna_distance)
plot(nj_tree, cex = .4)

# this is the most simplistic phylogenetic tree to be made and just shows the relation between the different samples.
# now lets change it up so it isnt as clutured and make it give the information we actually want

# we can also make it our tree rooted by adding the first sample ever taken

nj_rooted <- root(nj_tree, 1) %>% ladderize()

# performing a bootstrap will not only tell us how good our lines are, it also give us the option to collapse some of them

# we have to load the dna back with another method because otherwise boot.phylo cannot recognize it for some reason
dna_boots <- fasta2DNAbin(here("data/MERS_nucleotides.fa"))
bootstrap <- boot.phylo(nj_rooted, dna2, function(e) {root(bionj(dist.dna(e, model = "TN93")), 1)})

# because I am not really interested in the accession number of every sample i want to change it to the country name 

nj_rooted_renamed <- rename_taxa(nj_rooted, metadata, accession, country)

# now we can add our bootstrap values
plot(nj_rooted_renamed, cex = .4)
axisPhylo()
nodelabels(bootstrap, cex = .7)

    # now we can collapse some of our nodes
    nj_collapsed <- nj_rooted_renamed

    # step 1: figure out the row numbers of the branches that can be collapsed
    N <- length(nj_rooted_renamed$tip.label)
    tocollapse <- match(which(bootstrap<65)+N, nj_rooted_renamed$edge[,2])

    # step 2: get rid of those branches
    nj_collapsed$edge.length[tocollapse] <- 0
    nj_collapsed <- di2multi(nj_collapsed, tol = 0.00001)

    # step 3: check if it worked by plotting this new tree
    plot(nj_collapsed, cex = .45)

# lets add some nice colours to the plot so we can see the countries more easily
    
  plot(nj_collapsed, show.tip = FALSE)
  title("NJ tree for relation between samples for different countries")
  mycolours <- colorRampPalette(rep(c("blue", "green", "pink", "purple", "white", "red", "yellow", "black"), 2))
  tiplabels(nj_collapsed$tip.label, bg = transp(fac2col(as.factor(metadata$country), col.pal = mycolours), .3), cex = .4)
  countries <- metadata$country %>% unique()
  legend("bottomleft", fill = fac2col(as.factor(countries), col.pal = mycolours), legend = countries)





plot(nj_tree_renamed, cex = .4)


ggplot(nj_tree_renamed) +
  geom_tree() +
  geom_tiplab(size = 3)





```






```{r practise, eval=FALSE}
library(adegenet)
library(here)
library(tidyverse)
library(ggplot2)
#BiocManager::install("ggtree")
library(ggtree)




dna2 <- fasta2DNAbin(here("data/MERS_nucleotides.fa"))

dna2






dna <- fasta2DNAbin(file = here("data/MERS_nucleotides.fa"))
dna

annot <- read.csv(here("data/MERS_annotation.csv"))

library(ape)
D <- dist.dna(dna, model = "TN93")
length(D)

temp <- as.data.frame(as.matrix(D))
table.paint(temp, cleg=0, clabel.row=.5, clabel.col=.5)


tre <- nj(D)
class(tre)

h_cluster <- hclust(D, method = "average", members = NULL) # method = average is used for UPGMA, members can be equal to NULL or a vector with a length of size D
plot(h_cluster, cex = 0.6, show.tip = F)

tre <- ladderize(tre)

plot(tre, cex = 0.6)

annot <- annot %>% separate(collection_date, into = c("collection_date"), sep = 4)
annot$collection_date <- as.numeric(annot$collection_date)

plot(tre, show.tip=FALSE, main = "Unrooted NJ tree") # gets rid of the labels on the end, refer to the first tree depicted above
title("Unrooted NJ tree")
myPal <- colorRampPalette(c("red","yellow","green","blue"))
tiplabels(annot$collection_date, bg=num2col(annot$collection_date, col.pal=myPal), cex=.5) #we use the annot dataset to get our years
temp <- pretty(2013:2019, 2)
legend("bottomleft", fill=num2col(temp, col.pal=myPal), leg=temp, ncol=2)

tre2 <- root(tre, out = 1)
tre2 <- ladderize(tre2)

x <- as.vector(D)
y <- as.vector(as.dist(cophenetic(tre2)))
plot(x, y, xlab="original pairwise distances", ylab="pairwise distances on the tree", main="Is NJ appropriate?", pch=20, col=transp("black",.1), cex=3)
abline(lm(y~x), col="red")



tre3 <- as.phylo(hclust(D,method="average"))
y <- as.vector(as.dist(cophenetic(tre3)))
plot(x, y, xlab="original pairwise distances", ylab="pairwise distances on the tree", main="Is UPGMA appropriate?", pch=20, col=transp("black",.1), cex=3)
abline(lm(y~x), col="red")

cor(x, y)^2





myBoots <- boot.phylo(tre2, dna, function(e) root(nj(dist.dna(e, model = "TN93")),1))

myBoots


plot(tre2, show.tip=FALSE, edge.width=2)
title("NJ tree + bootstrap values")
tiplabels(frame="none", pch=20, col=transp(num2col(annot$collection_date, col.pal=myPal),.7), cex=3, fg="transparent")

axisPhylo()
temp <- pretty(2013:2019, 5)
legend("topright", fill=transp(num2col(temp, col.pal=myPal),.7), leg=temp, ncol=2)
nodelabels(myBoots, cex=.6)

```

